hyperopt/spark.py

```
@@ -0,0 +1,415 @@
+                            i.e., maximum number of concurrent Spark tasks.
+                            If set to None or and invalid value, this will be set to the number of
+                            executors in your Spark cluster.
+                            Hard cap at `MAX_CONCURRENT_JOBS_ALLOWED`.
+                            Default: None (= number of Spark executors).
+        :param timeout: Maximum time (in seconds) which fmin is allowed to take.
+                        If this timeout is hit, then fmin will cancel running and proposed trials.
+                        It will retain all completed trial runs and return the best result found
+                        so far.
+        :param spark_session: A SparkSession object. If None is passed, SparkTrials will attempt
+                              to use an existing SparkSession or create a new one. SparkSession is
+                              the entry point for various facilities provided by Spark. For more
+                              information, visit the documentation for PySpark.
+        """
+        super(SparkTrials, self).__init__(exp_key=None, refresh=False)
+        self._spark_context = SparkSession.builder.getOrCreate().sparkContext if spark_session is None \
+                                  else spark_session.sparkContext
+        # The feature to support controlling jobGroupIds is in SPARK-22340
+        self._spark_supports_job_cancelling = hasattr(self._spark_context.parallelize([1]),
+                                                      "collectWithJobGroup")
+        # maxNumConcurrentTasks() is a package private API
+        max_num_concurrent_tasks = self._spark_context._jsc.sc().maxNumConcurrentTasks()
+        self.parallelism = self._decide_parallelism(max_num_concurrent_tasks, parallelism)
+        self.user_specified_parallelism = parallelism
+
+        if timeout is not None and (not isinstance(timeout, numbers.Number) or timeout <= 0):
```

we can do the timeout check right after the super call to fail early.